lib/http/form_data/composite_io.rb

```diff
@@ -29,14 +29,16 @@ def initialize(ios)
       #
       # @return [String, nil]
       def read(length = nil, outbuf = nil)
-        outbuf = outbuf.to_s.replace("")
+        outbuf = outbuf.to_s.clear
```

Comment:
Though this is likely a bug in the previous version as well, `#to_s` may make a new object, at which point the purpose of passing `outbuf` is defeated.

Perhaps give outbuf a default value (e.g. `outbuf = "".b`), and raise if it's not a `String`?

Reply:
#to_s will not create a new object if the object is a string. But the replace call might change its encoding. That was the reasoning behind it.

Reply:
Hence "*may* make a new object". We *never* want to make a new object for the buffer... that defeats the point.

How about raising if the supplied buffer is not a `String` type with `Encoding::BINARY`?

Reply:
> We never want to make a new object for the buffer... that defeats the point.

ahh, I see what you mean. 

> How about raising if the supplied buffer is ...

I'm still not sure. The encoding requirement is usually not mandatory, and the buffer passed may be a specialized buffer. Besides "fixing" jruby, I'm not even sure that forcing binary encoding there is the right thing to do.

By the way, what's the reasoning behind allowing `length = nil`? The length argument is usually required for `IO#read` and `File#read`.

Reply:
> the buffer passed may be a specialized buffer

In what case would it make sense for it to be anything but a `String`? Ultimately it needs to thunk to a Ruby I/O function which also accepts a preallocated buffer, and those all need to be a `String` unless JRuby has some fancy behavior I don't know about where it lets you pass a Java ByteBuffer or something.

> Besides "fixing" jruby, I'm not even sure that forcing binary encoding there is the right thing to do.

For these types of buffers, I think mandating `Encoding::BINARY` is the only right option. They should really be a `ByteArray`/`ByteBuffer` type, but Ruby doesn't have those, so we need the next best thing, which is a `String` with `Encoding::BINARY`.

For other encodings there are all sorts of pathological performance implications, and the contract at this point in the code is to handle arbitrary binary data that may not be whatever-encoding-was-passed-in.

Given this is a rarely used power-user feature, I think it should have some pretty exacting semantics, which are:

- Avoid all allocations. The whole point of this API is to avoid allocations. This means no `#to_s` or anything else that might potentially allocate memory
- `String` is probably the only argument type that makes sense
- `Encoding::BINARY` (i.e. `Encoding::ASCII_8BIT`) is probably the only sensible encoding for the buffer `String`

> By the way, what's the reasoning behind allowing length = nil?

No idea, like the `outbuf` being `nil` I think that probably doesn't make sense.

Reply:
I agree with this change, I'll try to answer the questions below.